rclcpp/src/rclcpp/generic/generic_publisher.cpp

```diff
@@ -0,0 +1,59 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/generic/generic_publisher.hpp"
+
+#include <memory>
+#include <string>
+
+#include "rclcpp/generic/typesupport_helpers.hpp"
+
+namespace rclcpp
+{
+namespace generic
+{
+
+namespace
+{
+rcl_publisher_options_t get_publisher_options(const rclcpp::QoS & qos)
+{
+  auto options = rcl_publisher_get_default_options();
+  options.qos = qos.get_rmw_qos_profile();
+  return options;
+}
+}  // unnamed namespace
+
+GenericPublisher::GenericPublisher(
+  rclcpp::node_interfaces::NodeBaseInterface * node_base,
+  std::shared_ptr<rcpputils::SharedLibrary> ts_lib,
+  const std::string & topic_name,
+  const std::string & topic_type,
```

Comment:
passing in `ts_lib` and `topic_type` kind of feels redundant here to me. The type support library is loaded externally, why not passing in the type support handle with it directly?

Reply:
`topic_type` is still needed to construct the symbol name. Are you suggesting to change the signature of the GenericPublisher constructor to additionally take a type support handle (i. e. `rosidl_message_type_support_t *`)? Deriving that from `ts_lib` is cleaner imo. Or are you arguing that it should receive a type support handle instead of a `std::shared_ptr<rcpputils::SharedLibrary>`? That means the GenericPublisher couldn't be self-contained anymore, since the type support handle is a pointer into that library and the library needs to be kept alive at least as long as the publisher.

Reply:
Looking at this again, I think the API is okay as-is. I thought one could derive the actual topic type from the loaded typesupport library, but that might end up in some sort of a search.